### DISCUSSION TITLE
[v3] Javascript function run twice

### DISCUSSION THREAD

**Discussion by si-aji** [livewire/livewire#6515] [1. Help] [answered]
I have 2 Livewire component,
- Home
- Profile

I have a javascript function inside Profile page to generate random string and display it to element that exists in my layout. Here is some of my code

resources/views/components/layouts/app.blade.php
```php
...
@persist('data-attr')
     <div class=" fixed bottom-4 w-full items-center justify-center">
         <span class=" flex justify-center" data-attr="sample"></span>
     </div>
@endpersist
...
```

`resources/views/livewire/profile/index.blade.php`
```php
<div>
    {{-- Stop trying to control. --}}
    <h1 class=" text-3xl">Profile</h1>

    <div class=" mt-6">
        <button class=" px-4 py-2 rounded bg-blue-500 hover:bg-blue-700 transition-all text-white" id="profile-btn">Generate</button>
    </div>

    <script>
        function generateRandomString()
        {
            // Update DOM Element
            if(document.querySelector('[data-attr="sample"]')){
                console.log('Random String function (generateRandomString) are called');

                const randomStrings = Array.from({ length: 12 }, () =>
                    Math.random().toString(36).substring(2, 14)
                );

                const randomIndex = Math.floor(Math.random() * randomStrings.length);
                const randomString = randomStrings[randomIndex];
                document.querySelector('[data-attr="sample"]').innerHTML = randomString;
            }
        }

        document.addEventListener('livewire:navigated', () => {
            // Generate Random String on Init
            generateRandomString();
        });

        // Add button Event
        if(document.getElementById('profile-btn')){
            document.getElementById('profile-btn').addEventListener('click', () => {
                generateRandomString();
            });
        }
    </script>
</div>
```

There's 2 way to call `generateRandomString` function, first is when Profile page is visited with `livewire:navigated` or when the Generate button is clicked. Everything work as expected, string get generated when Profile page is navigated from Home, or when the button is clicked until I go back to Home page.

I don't write any function to call `generateRandomString` in home page, but it's seem the function still got called every time I navigate.

https://github.com/livewire/livewire/assets/31198572/90417e8f-76db-4a7d-9c67-d294a9b5a8b3

Here is my repo, [https://github.com/si-aji/LivewireV3](https://github.com/si-aji/LivewireV3). Is there something wrong there?

**Comment by si-aji** [comment, marked as answer]
Just asked ChatGPT according this. So 

```javascript
() => {
    // Generate Random String on Init
    generateRandomString();
}
```

is already binded to `livewire:navigated` event listener, so even when I navigate to other page, `generateRandomString` still got called because related function already binded to the `livewire:navigated` event listener.

So changing my function to this, do the job for me

```javascript
<script data-navigate-track>
    function generateRandomString()
    {
        // Update DOM Element
        if(document.querySelector('[data-attr="sample"]')){
            console.log('Random String function (generateRandomString) are called');

            const randomStrings = Array.from({ length: 12 }, () =>
                Math.random().toString(36).substring(2, 14)
            );

            const randomIndex = Math.floor(Math.random() * randomStrings.length);
            const randomString = randomStrings[randomIndex];
            document.querySelector('[data-attr="sample"]').innerHTML = randomString;
        }
    }

    function listener(){
        generateRandomString();
    };
    document.addEventListener('livewire:navigated', listener);
    document.addEventListener('livewire:navigating', () => {
        document.removeEventListener('livewire:navigated', listener);
    });

    // Add button Event
    if(document.getElementById('profile-btn')){
        document.getElementById('profile-btn').addEventListener('click', () => {
            // navigatedEventListener();
            console.log('Click via javascript');
        });
    }
</script>
```

Where when `livewire:navigated` is called, bind `listener` function and when navigating away (`livewire:navigating) remove the bind function

**Reply by janglapuk** [reply]
Have you put the script in the `<head>`?

> If two pages include the same `<script>` tag in the `<head>`, that script will only be run on the initial page visit and not on subsequent page visits.

- [Navigate](https://livewire.laravel.com/docs/navigate)
- [Scripts in `<head>` are loaded once](https://livewire.laravel.com/docs/navigate#scripts-in-head-are-loaded-once)

**Comment by vitalijalbu** [comment]
maybe you are inb dev mode, in react also js runs twice if dev

**Comment by mohammed2991991** [comment]
put data-navigate-once="true"
